packages/overlay/src/overlay-stack.ts

```diff
@@ -54,47 +83,37 @@ export class OverlayStack {
     }
 
     private isOverlayActive(overlayContent: HTMLElement): boolean {
-        return !!this.overlays.find((item) =>
-            overlayContent.isSameNode(item.overlayContent as HTMLElement)
-        );
+        return !!this.findOverlayForContent(overlayContent);
     }
 
     private isClickOverlayActiveForTrigger(trigger: HTMLElement): boolean {
-        return this.overlays.some(
-            (item) =>
-                trigger.isSameNode(item.trigger as HTMLElement) &&
-                item.interaction === 'click'
-        );
+        const overlay = this.findOverlayForTrigger(trigger);
+        return overlay != null && overlay.interaction === 'click';
     }
 
-    public openOverlay(event: CustomEvent<OverlayOpenDetail>): void {
-        if (this.isOverlayActive(event.detail.content)) return;
+    public openOverlay(details: OverlayOpenDetail): void {
+        /* istanbul ignore if */
+        if (this.isOverlayActive(details.content)) return;
 
         requestAnimationFrame(() => {
-            const interaction = event.detail.interaction;
-            if (interaction === 'click') {
+            if (details.interaction === 'click') {
                 this.closeAllHoverOverlays();
             } else if (
-                interaction === 'hover' &&
-                this.isClickOverlayActiveForTrigger(event.detail.trigger)
+                details.interaction === 'hover' &&
+                this.isClickOverlayActiveForTrigger(details.trigger)
             ) {
                 // Don't show a hover popover if the click popover is already active
                 return;
             }
 
-            const activeOverlay = ActiveOverlay.create(event, this.root);
-            this.overlays.push(activeOverlay);
-
-            this.onChange(this.overlays);
+            const activeOverlay = ActiveOverlay.create(details);
+            document.body.appendChild(activeOverlay);
```

Comment:
With the structural changes herein, if I wanted to make the "overlay root" element flexible, where would you see adding that? Having this as a `static` on `Overlay` makes me feel like it's a little fixed, but I don't use a lot of statics in my code to be fully aware of the possibilities they afford.

The issue this supports is something that is currently in Crisp where we don't actually use `sp-theme` to set our styles, as previously there was no way to use it with only _some_ of the styles. The answer may simply be for me to add the `large` styles to the `sp-theme` package, but an alternative would be to allow for our `crisp-app` element to be set for use in place of `document.body` here by sort sort of configuration.

Reply:
I too am a little concerned about not being able to control where we reparent overlays to.

The case I'm thinking of is a modal thats contained within an area of an application, in this case I want its z-index to be above the content in the containment area, but below content in other areas outside of the container. This seems natural to me to create an overlay root thats specifically for use by this modal, with the main document.body root being used for tooltips etc.

Is this a case this change supports? I think we may have even lost this capability before this change?

Reply:
This was a feature taken away when removing `overlay-root`, yes. I don't personally have a preference between needing it now vs the ability to easily refactor this functionality in later, which is why it was omitted previously when this wasn't a `static`.

@benjamind could you expand on:
> The case I'm thinking of is a modal thats contained within an area of an application, in this case I want its z-index to be above the content in the containment area, but below content in other areas outside of the container.

Is this a case of a UI with multiple modals open at the same time, or something else? Would be a good use case to have clearly defined in the docs as a "possible" or "not possible" depending on how this comes together.

Reply:
The case I was thinking of was something like this:

![image](https://user-images.githubusercontent.com/332843/72099889-6fe6ac80-32d6-11ea-987e-b39f01f11f79.png)

Where we have an area where we want a contained modal, this has a mask area behind the modal and a dialog with tooltips etc inside. This is one stack of overlays.

The application however supports dropdowns and tooltips that can potentially overlap this stack, but do so from outside of the contained area.

I've not delved too deeply into how our overlay stack works at this time, so I'm not sure how we support this today, especially with regards to resize of the masked area etc.

Reply:
So, our overlays stack globally in the order in which they are opened. I could see the dialog being an overlay itself. If you opened an overlay in the dialog (tooltip on hover, click popover, etc) it would open at a higher level than the dialog. If you opened a dropdown from the header, while the dialog was open (ignoring that it is supposed to be modal), the dropdown would appear over the dialog as you would expect.

A problem with explicitly restricting the overlay area to a subset of the view (e.g. the mask) is that we may restrict the positioning of something like a dropdown more than we need to. I don't know if that is something that we want?

So, I guess I am saying that I think we would get reasonable results in the scenario you have outlined without multiple levels of overlay root. I am not yet convinced that the added complexity is warranted. But, I could be convinced 😁

Reply:
Yeah I think I agree that our current code covers the case for the z-stacking issue.

The main concern I have is I think about responding to resize. Since the modal mask and the dialog would be centered within their container, but they are re-parented to `document.body` (or our overlay root), then I'm not quite sure how they would manage their resizing without having to have very explicit code to manage sizing the modal and dialog against their original container?